exo-go/src/service_restarter/index.go

```diff
@@ -0,0 +1,79 @@
+package serviceRestarter
+
+import (
+	"fmt"
+
+	"github.com/Originate/exosphere/exo-go/src/docker_compose"
+	"github.com/fsnotify/fsnotify"
+)
+
+// ServiceRestarter watches the given local service for changes and restarts it
+type ServiceRestarter struct {
+	ServiceName      string
+	ServiceDir       string
+	DockerComposeDir string
+	Env              []string
+	Log              func(string)
+	watcher          *fsnotify.Watcher
+}
+
+// Watch watches the service directory for changes
+func (s *ServiceRestarter) Watch(watcherErr chan<- error) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+	s.watcher = watcher
+	go func() {
+		for {
+			select {
+			case event := <-watcher.Events:
+				if isCreate(event) {
+					s.Log(fmt.Sprintf("Restarting service '%s' because %s was created", s.ServiceName, event.Name))
+				} else if isRemove(event) {
+					s.Log(fmt.Sprintf("Restarting service '%s' because %s was deleted", s.ServiceName, event.Name))
+				} else if isChange(event) {
+					s.Log(fmt.Sprintf("Restarting service '%s' because %s was changed", s.ServiceName, event.Name))
+				}
+				watcherErr <- s.restart(watcherErr)
+			case err := <-watcher.Errors:
+				watcherErr <- err
+			}
+		}
+	}()
+	return watcher.Add(s.ServiceDir)
+}
+
+func isChange(event fsnotify.Event) bool {
+	return (event.Op&fsnotify.Write == fsnotify.Write) || (event.Op&fsnotify.Rename == fsnotify.Rename) || (event.Op&fsnotify.Chmod == fsnotify.Chmod)
+}
+
+func isCreate(event fsnotify.Event) bool {
+	return event.Op&fsnotify.Create == fsnotify.Create
+}
+
+func isRemove(event fsnotify.Event) bool {
+	return event.Op&fsnotify.Remove == fsnotify.Remove
+}
+
+func (s *ServiceRestarter) restart(watcherErr chan<- error) error {
+	if err := s.watcher.Close(); err != nil {
+		return err
+	}
+	if err := dockerCompose.KillContainer(s.ServiceName, s.DockerComposeDir, s.Log); err != nil {
+		s.Log(fmt.Sprintf("Docker failed to kill container %s", s.ServiceName))
```

Comment:
Thoughts on wrapping the error instead of logging and leaving it the listener of the error channel to log? I believe currently there are some errors pushed into the channel that aren't wrapped and also not logged because the listener on the channel does not log the error passed to it.